tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java

```diff
@@ -185,23 +193,48 @@ private static boolean canBePruned(TiTableInfo tblInfo, Expression filter) {
     return pruneRangeNormalPart(tableInfo, cnfExpr);
   }
 
+  // say we have a partitioned table with the following partition definitions with year(y) as
+  // partition expression:
+  // 1. p0 less than 1995
+  // 2. p1 less than 1996
+  // 3. p2 less than maxvalue
+  // Above infos, after this function, will become the following:
+  // 1. p0: year(y) < 1995
+  // 2. p1: 1995 <= year(y) and year(y) < 1996
+  // 3. p2: 1996 <= year(y) and true
+  // true will become {@Code Constant} 1.
   private static List<Expression> generateRangePartExprs(TiTableInfo tableInfo) {
     TiPartitionInfo partInfo = tableInfo.getPartitionInfo();
     List<Expression> partExprs = new ArrayList<>();
     TiParser parser = new TiParser(tableInfo);
+    // check year expression
+    // rewrite filter condition
+    // purchased > '1995-10-10'
+    // year(purchased) > year('1995-10-10')
+    // purchased > 1995
     String partExprStr = tableInfo.getPartitionInfo().getExpr();
     partExpr = parser.parseExpression(partExprStr);
+    partExprColRefs = PredicateUtils.extractColumnRefFromExpression(partExpr);
     // when it is not range column case, only first element stores useful info.
     for (int i = 0; i < partInfo.getDefs().size(); i++) {
       TiPartitionDef pDef = partInfo.getDefs().get(i);
+      String current = pDef.getLessThan().get(0);
+      String leftHand;
+      if (current.equals("MAXVALUE")) {
+        leftHand = "true";
+      } else {
+        leftHand = String.format("%s < %s", partExprStr, current);
+      }
       if (i == 0) {
-        String literal = pDef.getLessThan().get(0);
-        partExprs.add(parser.parseExpression(partExprStr + "<" + literal));
+        partExprs.add(parser.parseExpression(leftHand));
       } else {
         String previous = partInfo.getDefs().get(i - 1).getLessThan().get(0);
-        String current = pDef.getLessThan().get(0);
-        String and =
-            String.format("%s and %s", partExprStr + ">=" + previous, partExprStr + "<" + current);
+        String and;
+        if (leftHand.equals("true")) {
```

Comment:
can't we use current.equals("MAXVALUE") directly here? what's the point to introduce "true"?

Reply:
We can't see "single maxvalue partition table case" test case.